File: src/data/types/crewMember.ts
import type { TierColor } from "./tierColor.ts";
import type { Profession } from "./profession.ts";
import type { ShipCrewRole } from "./shipCrewRole.ts";

// Phase 4 GDD §2/§3. Timestamps are epoch-ms numbers, matching this
// codebase's existing Date.now() convention (same as Phase 3's Listing).
export interface CrewMember {
  id: string;
  hiredByPlayerId: string;
  tier: TierColor;
  // null for tiers 3-5 (general/unspecialized); set and locked at hire
  // time for tiers 6-7 (§2.2) -- never reassigned after hiring.
  profession: Profession | null;
  status: "idle" | "active";
  assignedCraftId: string | null;
  hiredAt: number;
  // For background/idle catch-up resolution (§2.1a) -- elapsed time is
  // always derived as currentTime - lastCheckedAt, never caller-supplied.
  lastCheckedAt: number;
  wageAmount: number;
  // Used to detect unpaid-upkeep attrition (§2.7) -- a grace period is
  // measured from this timestamp, not from hiredAt.
  lastPaidAt: number;
  // Combat GDD §3: necessary correction to the amendment's own literal
  // pseudocode (`unavailableUntil: timestamp | null // new field only`,
  // written without a `?`) -- same backward-compatibility reasoning as
  // Voyage.isRetreat: a CrewMember persisted before this amendment
  // shipped has no such field at all, and this amendment's own testing
  // requirement demands existing prior-phase data still validates without
  // changes. Optional AND nullable: absent (pre-Combat data) and `null`
  // (available, post-Combat) mean the same thing -- "not currently
  // unavailable" -- set to a future epoch-ms timestamp only by a combat
  // loss (§2.5), cleared back to `null` once that time passes.
  unavailableUntil?: number | null;
  // Ship Crew Roles amendment. Optional/nullable, same backward-
  // compatibility shape as unavailableUntil -- a CrewMember persisted
  // before this amendment has neither field at all; missing and `null`
  // both mean "not assigned to any ship role." A ship-role assignment is
  // separate from status/assignedCraftId -- independent axes, per the
  // "must not gate the other 4 roles" decision (ship.md).
  shipRole?: ShipCrewRole | null;
  assignedShipId?: string | null;
}

File: unity/ProfitableCore/Schema/CrewMember.cs
namespace Profitable.Core.Schema;

// Ports src/data/types/crewMember.ts. Timestamps are epoch-ms (long),
// matching every other ported time value in this project.
public enum CrewStatus
{
    Idle,
    Active,
}

public class CrewMember
{
    public string Id { get; set; } = string.Empty;
    public string HiredByPlayerId { get; set; } = string.Empty;
    public TierColor Tier { get; set; }

    // Null for tiers 1-5 (general/unspecialized); set and locked at hire
    // time for tiers 6-7 -- never reassigned after hiring.
    public string? Profession { get; set; }

    public CrewStatus Status { get; set; }
    public string? AssignedCraftId { get; set; }
    public long HiredAt { get; set; }

    // For background/idle catch-up resolution -- elapsed time is always
    // derived as currentTime - LastCheckedAt, never caller-supplied.
    public long LastCheckedAt { get; set; }
    public double WageAmount { get; set; }

    // Used to detect unpaid-upkeep attrition -- a grace period is
    // measured from this timestamp, not from HiredAt.
    public long LastPaidAt { get; set; }

    // Combat amendment (Sub-Phase F, not yet ported) -- nullable and
    // defaults to null so pre-Combat data still validates unchanged, same
    // backward-compatibility shape the TypeScript source's own comment
    // documents. Set to a future epoch-ms timestamp only by a combat
    // loss, cleared back to null once that time passes.
    public long? UnavailableUntil { get; set; }

    // Ship Crew Roles amendment (Sub-Phase D, not yet ported) -- same
    // backward-compatibility shape as UnavailableUntil. A ship-role
    // assignment is independent of Status/AssignedCraftId.
    public ShipCrewRole? ShipRole { get; set; }
    public string? AssignedShipId { get; set; }
}
